src/bluetooth-fw/nimble/nimble_store.c

```diff
@@ -0,0 +1,268 @@
+/*
+ * Copyright 2025 Core Devices LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <bluetooth/bonding_sync.h>
+#include <bluetooth/gap_le_connect.h>
+#include <comm/bt_lock.h>
+#include <host/ble_hs.h>
+#include <string.h>
+#include <system/logging.h>
+#include <system/passert.h>
+#include <util/list.h>
+
+#include "nimble_type_conversions.h"
+
+#define KEY_SIZE 16
+
+#define BLE_FLAG_SECURE_CONNECTIONS 0x01
+
+typedef struct {
+  ListNode node;
+  struct ble_store_value_sec value_sec;
+} BleStoreValue;
+
+static BleStoreValue *s_value_secs;
+
+static bool prv_nimble_store_find_sec_cb(ListNode *node, void *data) {
+  BleStoreValue *s = (BleStoreValue *)node;
+  struct ble_store_key_sec *key_sec = (struct ble_store_key_sec *)data;
+
+  return ble_addr_cmp(&s->value_sec.peer_addr, &key_sec->peer_addr) == 0;
+}
+
+static BleStoreValue *prv_nimble_store_find_sec(const struct ble_store_key_sec *key_sec) {
+  if (!ble_addr_cmp(&key_sec->peer_addr, BLE_ADDR_ANY)) {
+    return (BleStoreValue *)list_get_at((ListNode *)s_value_secs, key_sec->idx);
+  } else if (key_sec->idx == 0) {
+    return (BleStoreValue *)list_find((ListNode *)s_value_secs, prv_nimble_store_find_sec_cb,
+                                      (void *)&key_sec->peer_addr);
+  }
+
+  return NULL;
+}
+
+static int prv_nimble_store_read_our_sec(const struct ble_store_key_sec *key_sec,
+                                         struct ble_store_value_sec *value_sec) {
+  int ret = 0;
+  BleStoreValue *s;
+
+  bt_lock();
+
+  s = prv_nimble_store_find_sec(key_sec);
+  if (s == NULL) {
+    ret = BLE_HS_ENOENT;
+    goto unlock;
+  }
+
+  memset(value_sec, 0, sizeof(*value_sec));
+
+  value_sec->peer_addr = key_sec->peer_addr;
+  value_sec->key_size = KEY_SIZE;
+
+  if (s->value_sec.ltk_present) {
+    value_sec->ediv = s->value_sec.ediv;
+    value_sec->rand_num = s->value_sec.rand_num;
+    value_sec->ltk_present = true;
+    memcpy(value_sec->ltk, s->value_sec.ltk, KEY_SIZE);
+  }
```

Comment:
why does this not just copy over the entire buffer, like `read_peer_sec` does?

if NimBLE might want to store other thigns here in the future that we might not support, should we assert if they're set?

Reply:
KEY_SIZE is actually 16, the maximum we can get. Pebble storage does not support any other size, NimBLE does. The current storage layer is a bit of a hack, it just preserves PebbleOS things as they were, but NimBLE does things slightly different.

Reply:
Note: key size is checked to be _correct_